service/controller/resource/deployment/update.go

```diff
@@ -73,49 +130,69 @@ func (r *Resource) newUpdateChange(ctx context.Context, obj, currentState, desir
+	// to check the state of all deployments before we can safely go ahead with
+	// the update procedure.
+	for _, d := range currentDeployments {
+		allReplicasUp := allNumbersEqual(d.Status.AvailableReplicas, d.Status.ReadyReplicas, d.Status.Replicas, d.Status.UpdatedReplicas)
+		if !allReplicasUp {
+			r.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("cannot update any deployment: deployment '%s' must have all replicas up", d.GetName()))
+			return nil, nil
+		}
+	}
+
+	// We select one deployment to be updated per reconciliation loop. Therefore
+	// we have to check its state on the version bundle level to see if a
+	// deployment is already up to date. We also check if there are any other
+	// changes on the pod specs. In case there are none, we check the next one.
+	// The first one not being up to date will be chosen to be updated next and
+	// the loop will be broken immediately.
+DeploymentsLoop:
+	for _, currentDeployment := range currentDeployments {
+		desiredDeployment, err := getDeploymentByName(desiredDeployments, currentDeployment.Name)
+		if IsNotFound(err) {
+			// NOTE that this case indicates we should remove the current deployment
+			// eventually.
+			r.logger.LogCtx(ctx, "level", "warning", "message", fmt.Sprintf("not updating deployment '%s': no desired deployment found", currentDeployment.GetName()))
+			continue
+		} else if err != nil {
+			return nil, microerror.Mask(err)
 		}
 
-		// We select one deployment to be updated per reconciliation loop. Therefore
-		// we have to check its state on the version bundle level to see if a
-		// deployment is already up to date. We also check if there are any other
-		// changes on the pod specs. In case there are none, we check the next one.
-		// The first one not being up to date will be chosen to be updated next and
-		// the loop will be broken immediatelly.
-		for _, currentDeployment := range currentDeployments {
-			desiredDeployment, err := getDeploymentByName(desiredDeployments, currentDeployment.Name)
-			if IsNotFound(err) {
-				// NOTE that this case indicates we should remove the current deployment
-				// eventually.
-				r.logger.LogCtx(ctx, "level", "warning", "message", fmt.Sprintf("not updating deployment '%s': no desired deployment found", currentDeployment.GetName()))
-				continue
-			} else if err != nil {
-				return nil, microerror.Mask(err)
-			}
+		if !isDeploymentModified(desiredDeployment, currentDeployment) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("not updating deployment '%s': no changes found", currentDeployment.GetName()))
+			continue
+		}
 
-			if !isDeploymentModified(desiredDeployment, currentDeployment) {
-				r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("not updating deployment '%s': no changes found", currentDeployment.GetName()))
-				continue
+		// If worker deployment, check that master does not have any prohibited states before updating the worker
+		if tcK8sClient != nil {
```

Comment:
Can this be `nil`? It seems like reconciliation should stop if the TC is unreachable.

Reply:
You're right, that was originally for testing. Now the deployment resource enforces the tenant cluster requirement